### ISSUE TITLE
Traing Trouble

### ISSUE THREAD

**Issue by landmars** [ultralytics/yolov5#4019]
## ❔Question
Hi, guys. I followed the guidlines to clone yolov5, install all the required packages and tried to train the coco dataset. While the error occured as below:
** UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule.**
I thought it might be my GPU-pytorch problem so I changed to CPU-pytorch. And it worked. But I need to use my GPU to acclerate the training. So could help to find what is the problem?

## Additional context
GPU_type: GTX 1660S
Python 3.8.0
CUDA: 11.1
Package                 Version
----------------------- ---------------------
absl-py                 0.13.0
cachetools              4.2.2
certifi                 2021.5.30
chardet                 4.0.0
click                   8.0.1
colorama                0.4.4
configparser            5.0.2
cycler                  0.10.0
docker-pycreds          0.4.0
gitdb                   4.0.7
GitPython               3.1.18
google-auth             1.32.1
google-auth-oauthlib    0.4.4
grpcio                  1.38.1
idna                    2.10
kiwisolver              1.3.1
Markdown                3.3.4
matplotlib              3.4.2
numpy                   1.21.0
oauthlib                3.1.1
olefile                 0.46
opencv-python           4.5.2.54
pandas                  1.2.5
pathtools               0.1.2
Pillow                  8.3.1
pip                     21.1.3
promise                 2.3
protobuf                3.17.3
psutil                  5.8.0
pyasn1                  0.4.8
pyasn1-modules          0.2.8
pyparsing               2.4.7
python-dateutil         2.8.1
pytz                    2021.1
PyYAML                  5.4.1
requests                2.25.1
requests-oauthlib       1.3.0
rsa                     4.7.2
scipy                   1.7.0
seaborn                 0.11.1
sentry-sdk              1.3.0
setuptools              52.0.0.post20210125
shortuuid               1.0.1
six                     1.16.0
smmap                   4.0.0
subprocess32            3.5.4
tensorboard             2.5.0
tensorboard-data-server 0.6.1
tensorboard-plugin-wit  1.8.0
thop                    0.0.31.post2005241907
torch                   1.8.1+cu111
torchaudio              0.8.1
torchvision             0.9.1+cu111
tqdm                    4.61.2
typing-extensions       3.10.0.0
urllib3                 1.25.11
wandb                   0.10.33
Werkzeug                2.0.1
wheel                   0.36.2
wincertstore            0.2

**Comment by glenn-jocher** [comment]
@landmars a warning is not an error. You have not shown any errors.

**Comment by landmars** [comment]
@glenn-jocher I'm sorry for my miss. But this warning is the only difference between GPU and CPU type. Besides the GPU type would get 0 on mAP as  below:
** Epoch   gpu_mem       box       obj       cls     total    labels  img_size
       9/9      3.6G       nan       nan       nan       nan       141       640: 100%|██| 8/8 [00:12<00:00,  1.51s/it]
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|█| 4/4 [00:02<00:00,  1.38
                 all        128          0          0          0          0          0 **
and that was training for coco. using command: python train.py --epochs 10
So, I wonder is that possible for the first train?
Thank you.

**Comment by glenn-jocher** [comment]
@landmars there are no known issues with training. I'd recommend you update torch to latest.

**Comment by landmars** [comment]
@glenn-jocher tried with pytorch==1.9.0. The P, R, mAP value still got 0 after 1000 epochs.

**Comment by glenn-jocher** [comment]
@landmars 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/

**Comment by landmars** [comment]
@glenn-jocher
Hi, I've fixed problem by using yolov5-5.0 zip instead of git clone. In the same envs, the git version somehow makes trouble with gradient. The download code zip seems fine.
Thank you for your answer.

**Comment by chenkang455** [comment]
> @glenn-jocher
> Hi, I've fixed problem by using yolov5-5.0 zip instead of git clone. In the same envs, the git version somehow makes trouble with gradient. The download code zip seems fine.
> Thank you for your answer.

Hello , i met the same problem as yours , could you tell the reason why the original version is not satisfied?

**Comment by glenn-jocher** [comment]
@chenkang455 👋 hi, thanks for letting us know about this problem with YOLOv5 🚀. We've created a few short guidelines below to help users provide what we need in order to get started investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

* ✅ **Minimal** – Use as little code as possible that still produces the same problem
* ✅ **Complete** – Provide **all** parts someone else needs to reproduce your problem in the question itself
* ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

In addition to the above requirements, for [Ultralytics](https://ultralytics.com/) to provide assistance your code should be:

* ✅ **Current** – Verify that your code is up-to-date with current GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been resolved by previous commits.
* ✅ **Unmodified** – Your problem must be reproducible without any modifications to the codebase in this repository. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all of the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) and providing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by landmars** [comment]
@chenkang455 
Hi, this troubles me too. I guess the only possible reason is my Windows and cuda11 couldn't match. Hence, I tried all the cuda 11.x. and they all gave mAP 0 when I use GPU. So I thought windows envs have trouble in using pytorch. This problem is too complex for me to handle.  So my solution is just downgrade cuda to 10.2, download yolov5-5.0 and reinstall pytorch.

**Comment by glenn-jocher** [comment]
@landmars @chenkang455 sometimes Windows/Anaconda result in 0 detections, but we're not sure of the source of the problem either.

Can you try to run YOLOv5 using pip env instead of anaconda?

**Comment by chenkang455** [comment]
> @landmars @chenkang455 sometimes Windows/Anaconda result in 0 detections, but we're not sure of the source of the problem either.
> 
> Can you try to run YOLOv5 using pip env instead of anaconda?

Yes ,  I changed the version of the cuda from 11 to 10.2 and it worked, but my computer shows that it supports cuda11 which confuses me a lot. And i run my YOLOv5 using pip env it worked

**Comment by glenn-jocher** [comment]
@chenkang455 well maybe Anaconda is the 100% source of the problem.

Can the pip env environment run with CUDA 11?

**Comment by chenkang455** [comment]
> @chenkang455 well maybe Anaconda is the 100% source of the problem.
> 
> Can the pip env environment run with CUDA 11?

NO , i changed the version from 11 to the 10.2 and it worked.

**Comment by glenn-jocher** [comment]
@chenkang455 ah, so even the pip env doesn't work well with CUDA 11. Ok understood.

**Comment by ves1120** [comment]
Had the same problem on GTX1650. Might be an issue with the 1600 series. I had the same setup on 2 computers (python 3.8, torch1.8.2+cu111, pip env) with different cards GTX1050 and GTX1650. Training was working on 1050 but not on the 1650 card. Inference works fine on cuda 11.1 on either card so I couldn't work out what was wrong.

Training works on 1650 card after downgrading to cuda 10.2

**Comment by glenn-jocher** [comment]
We also get faster results on V100, A100 with CUDA 10.2 than 11
